olp-cpp-sdk-core/src/thread/Continuation.cpp

```diff
@@ -0,0 +1,213 @@
+  // further, true means do nothing
+  bool ProcessAsyncTask(std::shared_ptr<ProcessorInternal> context_copy,
+                        std::pair<ContinuationImpl::AsyncTaskType,
+                                  ContinuationImpl::TaskType>& task) {
+    auto& current_task = task;
+    if (!current_task.first) {
+      // go to the next task in the queue to finilize the task
+      context_copy->tasks_.pop();
+      return false;
+    }
+
+    auto callback = [this, context_copy, &current_task](void* arg) mutable {
+      // limit lifetime of the context shared_ptr to the function scope, for
+      // extra safety in case user's async function stores callback somewhere
+      // for forever
+      auto context = std::move(context_copy);
+      // save the arg (the identity function)
+      context->last_output_ = current_task.second(arg);
+
+      // re-enter ProcessAsyncTask() in order to finilize the task
+      if (context->last_output_) {
+        ProcessNextTask(context);
+      }
+    };
+
+    auto func = std::move(current_task.first);
+    current_task.first = nullptr;
+    func(context_copy->LastOutput(), std::move(callback));
+    if (context_copy->public_execution_context_.Cancelled()) {
+      return false;
+    }
+
+    return true;
+  }
+
+  void FinishQueueExecution(const std::shared_ptr<ProcessorInternal>& context) {
+    if (context->final_callback_) {
+      // the final execution context, all tasks are done or the execution
+      // is cancelled
+      context->final_callback_(context->LastOutput(), context->IsCancelled());
+      context->final_callback_ = nullptr;
+    }
+  }
+
+  // the function is called iteratively for the same ExecutionContext until all
+  // tasks are processed
+  void ProcessNextTask(const std::shared_ptr<ProcessorInternal>& context) {
+    if (!context->tasks_.empty() && !context->IsCancelled()) {
+      // run the task
+      auto& task = context->tasks_.front();
+      if (ProcessAsyncTask(context, task)) {
+        return;
+      }
+    }
+
+    if (context->tasks_.empty() || context->IsCancelled()) {
+      FinishQueueExecution(context);
+    } else {
+      // schedule the next task for the execution
+      auto scheduler = context->task_scheduler_.lock();
+      if (scheduler) {
+        scheduler->ScheduleTask(CreateFollowingTask(context));
+      }
+    }
+  }
+
+  // creates a task for execution by TaskContinuation
+  Task CreateFollowingTask(const std::shared_ptr<ProcessorInternal>& context) {
+    return [this, context] { ProcessNextTask(context); };
+  }
+
+  // starts the first task in a chain
+  void Start() {
+    auto current_task_queue_element = CreateFollowingTask(processor_);
+
+    auto scheduler = task_scheduler_.lock();
+    if (scheduler) {
+      scheduler->ScheduleTask(std::move(current_task_queue_element));
+    }
+  }
+
+ private:
+  struct ProcessorInternal {
+    std::weak_ptr<olp::thread::TaskScheduler> task_scheduler_;
```

Comment:
Why does it need to be weak_ptr. Can it survive the TaskContinuation/Continuation?

Reply:
removed TaskScheduler from Processor and ProcessorInternal.